storage/pebble/blocks_provider.go

```diff
@@ -0,0 +1,130 @@
+package pebble
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-evm-gateway/storage"
+	"github.com/onflow/flow-go/fvm/evm/events"
+	evmTypes "github.com/onflow/flow-go/fvm/evm/types"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	gethCommon "github.com/onflow/go-ethereum/common"
+	"github.com/onflow/go-ethereum/eth/tracers"
+)
+
+type ResultsCollector struct{}
+
+func (s *ResultsCollector) StorageRegisterUpdates() map[flowGo.RegisterID]flowGo.RegisterValue {
+	return map[flowGo.RegisterID]flowGo.RegisterValue{}
+}
+
+var _ evmTypes.ReplayResultCollector = (*ResultsCollector)(nil)
+
+type BlocksProvider struct {
+	blocks             storage.BlockIndexer
+	chainID            flowGo.ChainID
+	tracer             *tracers.Tracer
+	latestBlockPayload *events.BlockEventPayload
+}
+
+var _ evmTypes.BlockSnapshotProvider = (*BlocksProvider)(nil)
```

Comment:
the existing `Blocks` in blocks.go could implement this

Reply:
Unfortunately that is not possible, in my opinion. Inside the `processEvents` of the `Engine` type, we want to first replay the received EVM block and all its transactions, and if no error or state changes occur, we'll then continue with updating the local state index, the blocks / transactions / receipts indices. Which means, that the `Blocks` index, does not yet have that current block which is still being processed and replayed.
That's why I have added the `OnBlockReceived` & `OnBlockExecuted` methods, to keep track of the latest block that we're processing in `BlocksProvider`. This will allow the `GetSnapshotAt` and `BlockContext` methods to work properly.